File: aoc/aoc2022day04campcleanup1.hs
-- uninhm
-- https://adventofcode.com/2022/day/4
-- implementation

{-# LANGUAGE OverloadedStrings #-}

import qualified Data.Text as T

data Range a = Range
  { start :: a
  , end :: a
  } deriving (Show, Eq)

fromList :: [a] -> Range a
fromList [s, e] = Range s e

contains :: Ord a => Range a -> Range a -> Bool
contains other range =
  start range <= start other && end other <= end range

parse :: String -> [[Range Int]]
parse = map parseLine . T.lines . T.pack
  where
    parseLine line =
      fromList . map (read . T.unpack) . T.splitOn "-" <$> T.splitOn "," line

solve :: [[Range Int]] -> Int
solve = length . filter id . map checkContaining
  where checkContaining [a, b] = contains a b || contains b a

main :: IO ()
main = print . solve . parse =<< readFile "input"

File: aoc/aoc2022day04campcleanup1.cpp
// vicfred
// https://adventofcode.com/2022/day/4
// implementation
#include <cstdio>
#include <fstream>
#include <iostream>

using namespace std;

int main() {
  ifstream input("input.txt");
  long long ans = 0LL;
  for (string line; getline(input, line);) {
    int v,w,x,y;
    sscanf(line.c_str(), "%d-%d,%d-%d", &v, &w, &x, &y);
    // cout << v << " " << w << " " << x << " " << y << endl;
    if(v <= x and y <= w or x <= v and w <= y) {
      ans += 1;
    }
  }
  cout << ans << endl;
  return 0;
}
